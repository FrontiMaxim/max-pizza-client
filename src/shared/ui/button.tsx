import { FC, ReactNode, SVGProps } from "react";
import cn from "classnames";

type ButtonProps = {
  variant?: "primary" | "secondary" | "default";
  disabled?: boolean;
  size?: "small" | "medium" | "large";
  children: ReactNode;
  onClick?: () => void;
  startIcon?: FC<SVGProps<SVGSVGElement>>;
  endIcon?: FC<SVGProps<SVGSVGElement>>;
  className?: string;
};

export const Button: FC<ButtonProps> = ({
  variant = "primary",
  children,
  onClick,
  startIcon,
  endIcon,
  size = "small",
  className,
  disabled,
}) => {
  const StartIcon = startIcon;
  const EndIcon = endIcon;

  return (
    <button
      onClick={onClick}
      className={cn(
        "font-semibold",
        "font-[16px]",
        "rounded-[15px]",
        "flex",
        "items-center",
        "transition duration-200 hover:shadow-lg",
        {
          ["px-[101px] py-[18px]"]: size === "large",
          ["px-[76px] py-[15px]"]: size === "medium",
          ["px-[18px] py-[11px]"]: size === "small",
          ["bg-primary text-white font-bold"]:
            variant === "primary" && !disabled,
          ["bg-secondary text-primary"]: variant === "secondary" && !disabled,
          ["bg-white text-primary border border-primary"]:
            variant === "default" && !disabled,
          ["bg-grey text-white"]: disabled,
        },
        className ?? ""
      )}
    >
      {StartIcon ? (
        <StartIcon
          className={cn("mr-[10px]", {
            ["stroke-white"]: variant === "primary" || disabled,
            ["stroke-primary"]:
              variant === "default" || (variant === "secondary" && !disabled),
          })}
        />
      ) : null}
      {children}
      {EndIcon ? (
        <EndIcon
          className={cn("ml-[10px]", {
            ["stroke-white"]: variant === "primary" || disabled,
            ["stroke-primary"]:
              variant === "default" || (variant === "secondary" && !disabled),
          })}
        />
      ) : null}
    </button>
  );
};
