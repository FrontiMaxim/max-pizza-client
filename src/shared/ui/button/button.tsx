import { FC, ReactNode, SVGProps } from "react";
import cn from "classnames";

type ButtonProps = {
  type?: "button" | "submit";
  variant?: "primary" | "secondary" | "default";
  disabled?: boolean;
  size?: "small" | "medium" | "large";
  children: ReactNode;
  onClick?: () => void;
  startIcon?: FC<SVGProps<SVGSVGElement>>;
  endIcon?: FC<SVGProps<SVGSVGElement>>;
  className?: string;
};

export const Button: FC<ButtonProps> = ({
  type = "button",
  variant = "primary",
  children,
  onClick,
  startIcon,
  endIcon,
  size = "small",
  className,
  disabled,
}) => {
  const StartIcon = startIcon;
  const EndIcon = endIcon;

  return (
    <button
      type={type}
      onClick={onClick}
      className={cn(
        "font-semibold rounded-[15px] flex items-center justify-center",
        {
          ["px-[101px] py-[18px]"]: size === "large",
          ["px-[76px] py-[15px]"]: size === "medium",
          ["px-[18px] py-[11px]"]: size === "small",
          ["bg-primary text-white font-bold transition duration-200 hover:shadow-lg"]:
            variant === "primary" && !disabled,
          ["bg-secondary text-primary transition duration-200 hover:shadow-lg"]:
            variant === "secondary" && !disabled,
          ["bg-white text-primary border border-primary transition duration-200 hover:shadow-lg"]:
            variant === "default" && !disabled,
          ["bg-grey text-white"]: disabled,
        },
        className ?? ""
      )}
      disabled={disabled}
    >
      {StartIcon ? (
        <StartIcon
          className={cn("mr-[10px]", {
            ["stroke-white"]: variant === "primary" || disabled,
            ["stroke-primary"]:
              variant === "default" || (variant === "secondary" && !disabled),
          })}
        />
      ) : null}
      {children}
      {EndIcon ? (
        <EndIcon
          className={cn("ml-[10px]", {
            ["stroke-white"]: variant === "primary" || disabled,
            ["stroke-primary"]:
              variant === "default" || (variant === "secondary" && !disabled),
          })}
        />
      ) : null}
    </button>
  );
};
