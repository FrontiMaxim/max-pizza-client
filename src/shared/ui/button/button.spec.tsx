import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { Button } from "./button";
import { vi } from "vitest";

describe("Button", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("shound button is render", async () => {
    render(<Button>button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toBeVisible();
  });

  it("should content is 'button'", async () => {
    render(<Button>button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement.textContent).toBe("button");
  });

  it("should focused", async () => {
    render(<Button>button</Button>);

    const buttonElement = await screen.findByRole("button");
    await userEvent.click(buttonElement);

    expect(buttonElement).toHaveFocus();
  });

  it("should variant is primary", async () => {
    render(<Button variant="primary">button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement).toMatchSnapshot();
  });

  it("should variant is default", async () => {
    render(<Button variant="default">button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement).toMatchSnapshot();
  });

  it("should variant is secondary", async () => {
    render(<Button variant="secondary">button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement).toMatchSnapshot();
  });

  it("should state is disabled ", async () => {
    render(<Button disabled>button</Button>);

    const buttonElement = await screen.findByRole("button");

    expect(buttonElement).toMatchSnapshot();
  });

  it("should handler click is called ", async () => {
    const mockHandleClick = vi.fn();

    render(<Button onClick={mockHandleClick}>button</Button>);

    const buttonElement = await screen.findByRole("button");

    await userEvent.click(buttonElement);

    expect(mockHandleClick).toHaveBeenCalledTimes(1);
  });

  it("should handler click returned value", async () => {
    const mockHandleClick = vi.fn().mockImplementation(() => 1);

    render(<Button onClick={mockHandleClick}>button</Button>);

    const buttonElement = await screen.findByRole("button");

    await userEvent.click(buttonElement);

    expect(mockHandleClick).toHaveReturnedWith(1);
  });

  it("should not click when state is disable", async () => {
    const mockHandleClick = vi.fn();

    render(
      <Button onClick={mockHandleClick} disabled>
        button
      </Button>
    );

    const buttonElement = await screen.findByRole("button");

    await userEvent.click(buttonElement);

    expect(mockHandleClick).not.toHaveBeenCalled();
  });
});
