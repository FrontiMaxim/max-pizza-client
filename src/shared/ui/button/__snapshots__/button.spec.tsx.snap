// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should state is disabled  1`] = `
<button
  class="font-semibold rounded-[15px] flex items-center justify-center px-[18px] py-[11px] bg-grey text-white"
  disabled=""
  type="button"
>
  button
</button>
`;

exports[`Button > should variant is default 1`] = `
<button
  class="font-semibold rounded-[15px] flex items-center justify-center px-[18px] py-[11px] bg-white text-primary border border-primary transition duration-200 hover:shadow-lg"
  type="button"
>
  button
</button>
`;

exports[`Button > should variant is primary 1`] = `
<button
  class="font-semibold rounded-[15px] flex items-center justify-center px-[18px] py-[11px] bg-primary text-white font-bold transition duration-200 hover:shadow-lg"
  type="button"
>
  button
</button>
`;

exports[`Button > should variant is secondary 1`] = `
<button
  class="font-semibold rounded-[15px] flex items-center justify-center px-[18px] py-[11px] bg-secondary text-primary transition duration-200 hover:shadow-lg"
  type="button"
>
  button
</button>
`;
